### PR TITLE
Storage engine config cleanup

### LINKED ISSUE #955: Unify storage engine configuration
We need to unify the storage engine configuration. The result could look like this on the config file:

```toml
[storage]
   ...
   ...
   ...

   [index]
       ...
       ...
```

We need to merge the `tsdb.Config` into the `storage.Config`, and remove the need to inject the `tsdb.Config` into the `tsm1.EngineOptions` struct.

### PR DESCRIPTION
**Pull request by zeebo**
This commit cleans up the configuration between the storage and tsm1 packages.

- The WAL config was moved up to storage.
- All relevant engine config was moved into the tsm1.Config.
- Many unused parameters and dead code were removed.
- Moved constants and defaults around to be closer to their owners.
- Added a `storage/compat` package to help migrate configs in current toml files to the new config layout.

One of these commits (the compat one) was authored by @jessuvius. Thanks!

Closes #955

### CODE REVIEW

File: storage/compat/compat.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package compat
```
Comment (e-dard) [inline storage/compat/compat.go]: nit: if `compat` is a new package, maybe a package level comment explaining the reason for its existence would be useful.

File: storage/config.go [hunk 2]
```diff
@@ -1,30 +1,91 @@
...
+		ValidateKeys:        DefaultValidateKeys,
+		TraceLoggingEnabled: DefaultTraceLoggingEnabled,
+
+		WAL:    tsm1.NewWALConfig(),
+		Engine: tsm1.NewConfig(),
+		Index:  tsi1.NewConfig(),
+	}
+}
+
+const (
```
Comment (e-dard) [inline storage/config.go]: Can these maybe go up to the top of the file?

File: tsdb/tsm1/config.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	// ThroughputBurst is the rate limit in bytes per second that we will allow TSM compactions
+	// to write to disk. If this is not set, the burst value will be set to equal the normal
+	// throughput
+	ThroughputBurst toml.Size `toml:"throughput-burst"`
+
+	// MaxConcurrent is the maximum number of concurrent full and level compactions that can
+	// run at one time.  A value of 0 results in 50% of runtime.GOMAXPROCS(0) used at runtime.
+	MaxConcurrent int `toml:"max-concurrent"`
+
+	// TODO(jeff): document
```
Comment (e-dard) [inline tsdb/tsm1/config.go]: Do you need some help with this?

File: tsdb/tsm1/engine.go [hunk 4]
```diff
@@ -28,17 +28,14 @@ import (
...
+//go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@iterator.gen.go.tmpldata iterator.gen.go.tmpl array_cursor.gen.go.tmpl array_cursor_iterator.gen.go.tmpl
 //go:generate env GO111MODULE=on go run github.com/influxdata/platform/tools/tmpl -i -data=file_store.gen.go.tmpldata file_store.gen.go.tmpl=file_store.gen.go
 //go:generate env GO111MODULE=on go run github.com/influxdata/platform/tools/tmpl -i -d isArray=y -data=file_store.gen.go.tmpldata file_store.gen.go.tmpl=file_store_array.gen.go
 //go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@encoding.gen.go.tmpldata encoding.gen.go.tmpl
 //go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@compact.gen.go.tmpldata compact.gen.go.tmpl
 //go:generate env GO111MODULE=on go run github.com/benbjohnson/tmpl -data=@reader.gen.go.tmpldata reader.gen.go.tmpl
 
-var (
-	// Ensure Engine implements the interface.
-	_ tsdb.Engine = &Engine{}
```
Comment (e-dard) [inline tsdb/tsm1/engine.go]: 🎉

File: tsdb/tsm1/config.go (hunk 3 above)
Comment (zeebo) [reply tsdb/tsm1/config.go]: Nope, just missed it.

**Comment by jacobmarble** [comment]
Thanks @jessuvius and @zeebo, this is great.

**Pull request merged**